src/rust/catnap/mod.rs

```diff
@@ -181,281 +174,266 @@ impl CatnapLibOS {
+                self.insert_coroutine(task_id, coroutine)
+            })?)
+    }
+
+    /// Asynchronous code to close a queue. This function returns a coroutine that runs asynchronously to close a queue
+    /// and the underlying POSIX socket and performs any necessary mult-queue operations at the libOS-level after
+    /// the close succeeds or fails.
+    fn close_coroutine(&self, qd: QDesc, yielder: Yielder) -> Result<Pin<Box<Operation>>, Fail> {
+        let qtable: Rc<RefCell<IoQueueTable<CatnapQueue>>> = self.qtable.clone();
+        let queue: CatnapQueue = self.get_queue(qd)?;
+
+        Ok(Box::pin(async move {
             // Wait for close operation to complete.
-            match queue.async_close(yielder).await {
+            match queue.do_close(yielder).await {
                 Ok(()) => {
-                    queue.commit();
-                    // Update socket.
-                    let mut qtable_: RefMut<IoQueueTable<CatnapQueue>> = qtable_ptr.borrow_mut();
-                    match qtable_.get_mut(&qd) {
-                        Some(queue) => {
-                            // Cancel all pending operations.
-                            queue.cancel_pending_ops(Fail::new(libc::ECANCELED, "This queue was closed"));
-                        },
-                        None => {
-                            let cause: &String = &format!("invalid queue descriptor: {:?}", qd);
-                            error!("{}", &cause);
-                            return (qd, OperationResult::Failed(Fail::new(libc::EBADF, cause)));
-                        },
-                    }
-                    queue.commit();
-                    qtable_.free(&qd);
+                    // Remove the queue from the queue table.
+                    qtable.borrow_mut().free(&qd);
                     (qd, OperationResult::Close)
                 },
                 Err(e) => {
-                    queue.abort();
                     warn!("async_close() qd={:?}: {:?}", qd, &e);
                     (qd, OperationResult::Failed(e))
                 },
             }
-        });
-        {
-            let task_id: String = format!("Catnap::close for qd={:?}", qd);
-            let task: OperationTask = OperationTask::new(task_id, coroutine);
-            let queue: CatnapQueue = self.get_queue(qd)?;
-
-            let handle: TaskHandle = match self.runtime.scheduler.insert(task) {
-                Some(handle) => {
-                    queue.commit();
-                    handle
-                },
-                None => {
-                    queue.abort();
-                    return Err(Fail::new(libc::EAGAIN, "cannot schedule co-routine"));
-                },
-            };
-            Ok(handle.get_task_id().into())
-        }
+        }))
     }
 
-    /// Pushes a scatter-gather array to a socket.
+    /// Synchronous code to pushto [buf] to a CatnapQueue and its underlying POSIX socket. This function schedules the
```

Comment:
You mean: `Synchronous code to push`? (not pushto)

Also, in the comment, do you mean to call it `buf` or `sga`? Applicable in other places as well. I think the best is to not mention it in the comment at all.